Admin: Add Google SSO for Compiler users

machikoyasuda opened the pull request:
closes #1817 

After merging this PR and ensuring the proper secrets are set up in Terraform, anyone with a `compiler.la` email address will be able to log into the Benefits Admin application. When the user is authenticated, the application will create a user with their Google first name, last name and e-mail address in Benefits. 

## What this PR does
- Uses https://megalus.github.io/django-google-sso/ library for Google single-sign on. Also uses https://github.com/encode/httpx/ for simple HTTP requests, to fetch Google API data.
- Set up Google SSO log in to allow for `compiler.la` domains, if the users email address is in the correct secret (see below)
- Set up Google SSO secrets with Terraform
- Fixes issue where Django JS files were not being allowed on Admin site
- Set user's username as their Google email address, and also save their profile icon URL, first and last name from Google.
- Outside of this PR, the necessary secrets have been set up. These secrets allow us to control exactly which `compiler.la` addresses can log in and get staff or superuser status:

```
GOOGLE_SSO_STAFF_LIST - andy@compiler.la
GOOGLE_SSO_SUPERUSER_LIST - angela@compiler.la,kegan@compiler.la,machiko@compiler.la
```



<img width="559" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/ec7cfe31-bdb9-4946-a407-b6d2394067fb">
<img width="503" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/dc54cf0d-89b4-49c9-b38f-1ee5522a1527">

## Done
- [x] Set up Google Cloud SSO new project for Benefits Admin
- [x] Set up OAuth page
- [x] Set up OAuth credentials per environment (Localhost, Dev, Test, Prod). Put in LastPass.
- [x] Set up Terraform with info from above
- [x] Fix test failures
- [x] Host Google image locally
- [x] Make Allowed Domain (`compiler.la, mst.org`) an environment variable (similar to Allowed Hosts - split by comma and filter empties)
- [x] Create username as your Google email address, save user's first name, last name and email address.
- [x] Create Allowable Staff list + env var
- [x] Create Allowable Admin list + env var
- [x] Add spec



## About `django-google-sso`
- Supports Django 5.0 [https://github.com/megalus/django-google-sso/releases](https://github.com/megalus/django-google-sso/releases/tag/v5.0.0)
- Supports Python 3.11 https://megalus.github.io/django-google-sso/#install
- Has a sibling Microsoft and GitHub library: https://megalus.github.io/django-google-sso/multiple/

## How to test
- Engineers: Try to log in via your Compiler email address. Should be able to log in successfully.
- Engineers: Try to log in via a personal Gmail address. Should not be able to log in.

machikoyasuda commented on benefits/settings.py:
```diff
@@ -281,7 +288,11 @@ def _filter_empty(ls):
+CSP_IMG_SRC = [
+    "'self'",
+    "data:",
+    "https://upload.wikimedia.org/wikipedia/commons/thumb/c/c1/Google_%22G%22_logo.svg/1280px-Google_%22G%22_logo.svg.png",
```
Just added this here for now, so it'll work out of the box. But we can upload our own Google logo and require it like this: https://megalus.github.io/django-google-sso/customize/#the-button-logo

machikoyasuda replied:
https://github.com/cal-itp/benefits/pull/1855/commits/8df45987b0d3f4e79e1cd182f1250e57b0bf4448

machikoyasuda commented on benefits/settings.py:
```diff
@@ -29,13 +29,20 @@ def _filter_empty(ls):
+    GOOGLE_SSO_PROJECT_ID = os.environ.get("GOOGLE_SSO_PROJECT_ID", "benefits-admin")
+    GOOGLE_SSO_CLIENT_SECRET = os.environ.get("GOOGLE_SSO_CLIENT_SECRET", "secret")
+
+    GOOGLE_SSO_ALLOWABLE_DOMAINS = ["compiler.la"]
```
@thekaveman Is it okay if the app allows _any_ user with the domain `compiler.la`? Or should it be a specific list of email addresses?

machikoyasuda replied:
Other possible options, depending on what we want:
- Set a certain list of email addresses to be created as superadmin, versus others are just staff permission levels: https://megalus.github.io/django-google-sso/users/#giving-permissions-to-auto-created-users

machikoyasuda replied:
@thekaveman Should this also be an env variable? (List of email addresses that would be superadmins, and everyone else with a compiler email address would be staff).

thekaveman replied:
Yeah I think we should start with a specific list of emails (from Compiler) that are allowed in, and ALSO a specific list that gets added as superusers (devs).

machikoyasuda replied:
In this case, I'd recommend we seed the database (save a copy of the DB with the user data) with this pre-configured list of users.

thekaveman replied:
We would need some kind of configuration model to be able to put this in the database then, right? Unless the SSO package provides a place for that to be stored?

A site configuration model is something I've been thinking about off and on. But we haven't exactly scoped that or talked about it as a team. So for now, I think an env var that reads from Key Vault via Terraform is the way to go.

machikoyasuda commented:
To test this locally so far:

- I used Google SSO test credentials from a personal app I created long, long ago. I can DM you the credentials if you want. And I replaced those lines here https://github.com/cal-itp/benefits/pull/1855/files#diff-d71405aa44f1e899a9cd920e5187aa1d02affdbd0da472eca95f814b0a8c1210R40-R42

```
    # GOOGLE_SSO_CLIENT_ID = "XXXX"
    # GOOGLE_SSO_PROJECT_ID = "XXXX"
    # GOOGLE_SSO_CLIENT_SECRET = "XXXX"
```

- Recreate devcontainer. Make sure it downloaded the Django-Google-SSO library. Running a `./bin/init.sh` runs `migrate` - and that should work successfully.
- Go to localhost:8000/admin
- Try logging in with a Compiler and a non-Compiler Google account. Should be able to log in and not log in, respectively.

machikoyasuda commented:
Not sure why GitHub Actions pytest/Cypress are failing here: https://github.com/cal-itp/benefits/actions/runs/7549568812/job/20553782925?pr=1855#step:7:22

I am able to run `migrate` locally and run all the tests locally, too.

cc @angela-tran @thekaveman

angela-tran commented:
@machikoyasuda I think if you were to run locally with `DJANGO_ADMIN` set to false, you would see the same issue when running migrations.

This can probably be resolved by moving [line 32](https://github.com/cal-itp/benefits/pull/1855/files#diff-d71405aa44f1e899a9cd920e5187aa1d02affdbd0da472eca95f814b0a8c1210R32) to go after line 49.

machikoyasuda commented:
@thekaveman @angela-tran Google SSO credential information for Localhost, Dev, Test and Prod are now saved in LastPass, Shared-Compiler Engineering.

machikoyasuda commented:
Terraform updated for 4 Google SSO env vars for Dev, Test and Prod.

thekaveman commented:
@machikoyasuda I'm taking a look at this now. One quick question, you mentioned:

> Also uses https://github.com/encode/httpx/ for simple HTTP requests

Is it possible to use `requests` here, which was already listed as a requirement? Rather than adding a new package that essentially does the same thing?

thekaveman reviewed (commented):
This is looking *fantastic* so far!! :clap: :clap: 

## Steps I took for local setup

### Created the following environment variables in my local `.env` file:

```env
DJANGO_ADMIN=true
DJANGO_SUPERUSER_EMAIL=superuser@calitp.org
DJANGO_SUPERUSER_PASSWORD=password1234!
DJANGO_SUPERUSER_USERNAME=benefits-admin

DJANGO_LOCAL_PORT=11369

GOOGLE_SSO_CLIENT_ID=<value from https://console.cloud.google.com/apis/credentials/oauthclient/415252694921-b03mul9832970e54gbkj4ae8l4f6fu4q.apps.googleusercontent.com?project=benefits-admin-411518>
GOOGLE_SSO_PROJECT_ID=benefits-admin-411518
GOOGLE_SSO_CLIENT_SECRET=<https://console.cloud.google.com/apis/credentials/oauthclient/415252694921-b03mul9832970e54gbkj4ae8l4f6fu4q.apps.googleusercontent.com?project=benefits-admin-411518>
GOOGLE_SSO_ALLOWABLE_DOMAINS=compiler.la
GOOGLE_SSO_STAFF_LIST=bot@compiler.la
GOOGLE_SSO_SUPERUSER_LIST=kegan@compiler.la
```

### Rebuild and Reopen devcontainer

Confirm in log output that `Admin` is enabled and setup:

```
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, core, django_google_sso, sessions
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying admin.0003_logentry_add_action_flag_choices... OK
...
```

### Login with the standard superuser

:ballot_box_with_check: Using the username/password as shown above

### Login with my personal compiler.la Google account

:ballot_box_with_check: A user was created for my Google account
:ballot_box_with_check: That user is marked as a superuser

### Login with the bot@compiler.la Google account

:ballot_box_with_check: A user was created for this Google account
:ballot_box_with_check: That user is marked as a normal staff user

### Attempt login with a non-compiler.la Google account

:ballot_box_with_check: Access is blocked due to non-matching domain

## A couple of questions

* [x] [Asked in another comment](https://github.com/cal-itp/benefits/pull/1855#issuecomment-1917582752): do we need to use `httpx` here when we already have `requests`?
* [x] Do we need the user's Google profile pic? I don't see that showing up anywhere in the Admin interface. Should I?
* [x] Slightly off-topic for this PR, but related: I think we can remove `127.0.0.1` from the default `ALLOWED_HOSTS` list -- since the Google integration is setup for `localhost` redirects only, we can simplify that setting a bit.
* [x] ~Can we get some quick unit tests written for `benefits/admin.py`, the `pre_login_user` function?~ Skip these for now, it's complicated to test models/database type stuff.

machikoyasuda commented:
@thekaveman Re: The icon - shows up here: `http://localhost:11369/admin/django_google_sso/googlessouser/1/change/`

<img width="822" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/6fa81ade-0254-4ad0-9054-8dac95db8eb3">

and the URL is used here: `http://localhost:11369/admin/auth/user/1/change/`

<img width="1088" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/ee328128-12b9-425f-8bd6-61f3593fef94">

machikoyasuda commented on benefits/admin.py:
```diff
@@ -0,0 +1,25 @@
+
+        # Request Google user info to get name and email
+        url = "https://www.googleapis.com/oauth2/v3/userinfo"
+        response = requests.get(url, headers=headers, timeout=settings.REQUESTS_TIMEOUT)
```
@thekaveman What should I set the timeout to? I just went with the app default here.

Tests are failing: https://github.com/cal-itp/benefits/actions/runs/7717803049/job/21037685917#step:4:181

thekaveman replied:
@machikoyasuda I think that test failure is the same one I was seeing yesterday related to the fix from #1862.

If you rebase on `dev` the tests should pass again? Your usage of the app default looks correct to me.

thekaveman commented:
@machikoyasuda any updates on adding some basic unit tests here?

machikoyasuda commented on tests/pytest/conftest.py:
```diff
@@ -3,14 +3,17 @@
-from pytest_socket import disable_socket
+
+# from pytest_socket import disable_socket
 
 from benefits.core import session
 from benefits.core.models import AuthProvider, EligibilityType, EligibilityVerifier, PaymentProcessor, PemData, TransitAgency
+from django.contrib.auth.models import User
 
 
 def pytest_runtest_setup():
-    disable_socket()
+    # disable_socket()
+    pass
```
temporary for local unit test running

thekaveman replied:
Probably related to this? https://github.com/microsoft/vscode-python/issues/22383

I added a workaround mentioned in that thread to my personal VSCode `settings.json`: https://github.com/thekaveman/dotfiles/commit/453a265767b5fd2481cd6dedad118c2f3cc608c9

machikoyasuda commented on tests/pytest/test_admin.py:
```diff
@@ -0,0 +1,31 @@
+import pytest
```
@thekaveman here's what i have so far:

- make an AdminUser fixture
- mocking `token` and _trying_ to mock the Google SSO response and json object - but the mock/patch isn't returning the values I want.

Basically trying to test: That before the `pre_login_user` method, a user has a blank first and last name. After the method runs, the user's first and last name come from the Google SSO response json object. Should I be taking a different approach? (Like just making sure the `pre_login_user` method gets called when a user logs in.. or something like that?)

machikoyasuda replied:
@thekaveman Good news - after refactoring the tests to use the `mocker` pattern, the tests are now passing as expected 👍 At least locally.

Now getting GitHub Action CI failures: https://github.com/cal-itp/benefits/actions/runs/7793576156/job/21253511521#step:8:50 

Not really sure why/how?? https://stackoverflow.com/questions/40206569/django-model-doesnt-declare-an-explicit-app-label

thekaveman replied:
Arg, I'm running into a similar issue with that same error, trying to test a custom field I'm creating for #1847.

I've been trying to find a solution for about the last hour but nothing yet :angry:

thekaveman replied:
I think at this point with all these problems running unit tests that need database/model access, we can probably just call this PR good to go. If you can just interactive rebase to drop the 3 commits related to testing [660f269](https://github.com/cal-itp/benefits/pull/1855/commits/660f26997a7e169cbbcaf6c4d2e03904b8d3a8e0), [0d5984e](https://github.com/cal-itp/benefits/pull/1855/commits/0d5984ef1b2c47321b13a4f0709989795c501723), and [2e28ec7](https://github.com/cal-itp/benefits/pull/1855/commits/2e28ec7de000169eba060b332e56519e2046547c), and then [move the `admin.py` file as you suggested](https://github.com/cal-itp/benefits/pull/1855#discussion_r1479147581), this can be approved for merge.

angela-tran replied:
The problem in that GitHub Action failure is once again because some apps are installed [only if `ADMIN` is enabled](https://github.com/cal-itp/benefits/blob/dev/benefits/settings.py#L44). 

We decided that the ultimate change we want to make regarding `ADMIN` is to remove it entirely (#1856), but it seems like if we went ahead and just removed this one `ADMIN` check, i.e. install all the apps, then the unit test @machikoyasuda wrote will work and hopefully whatever issue @thekaveman  is seeing in #1847 will also go away.

thekaveman replied:
My issue was unrelated to the `ADMIN` setting, it was in trying to unit test a custom field by attaching it to a model that doesn't exist in our migrations (just a model I defined directly in the tests).

@machikoyasuda timebox to about 30 minutes here, if you can get it working with minimal impact/changes then great, otherwise don't worry about the tests for now. Let's get this wrapped up.

angela-tran reviewed (commented):
I was able to test out setting all the environment variables up and logging in with my Compiler Google account today. I got through successfully too 👍

machikoyasuda commented on benefits/admin.py:
```diff
@@ -0,0 +1,25 @@
+        user.last_name = user_data["family_name"]
+        user.username = user_data["email"]
+        user.email = user_data["email"]
+        user.save()
```
@thekaveman @angela-tran Should this file be in/combined with what is already in `core/admin.py`?

thekaveman replied:
:+1: Yes I think so, no need to create an additional `admin.py` file at the top-level.

thekaveman commented:
I'm looking at this again now :eyes:

thekaveman commented on benefits/settings.py:
```diff
@@ -22,7 +22,7 @@ def _filter_empty(ls):
 
 ADMIN = os.environ.get("DJANGO_ADMIN", "False").lower() == "true"
```
I don't see any other references to this setting, so I guess we should delete it too?

angela-tran replied:
My suggestion was to do the minimal amount of change to get the test passing. The full removal of `DJANGO_ADMIN`  would be done in #1856

thekaveman commented on benefits/admin.py:
```diff
@@ -0,0 +1,25 @@
+from loguru import logger
+
+
+def pre_login_user(user, request):
```
This is now a duplicate of what is in `benefits/core/admin.py`? The entire file should probably be deleted right?

angela-tran replied:
I agree with this. I hadn't realized the file wasn't deleted

machikoyasuda commented:
Terraform dev has been updated to include:

```
GOOGLE_SSO_STAFF_LIST - x
GOOGLE_SSO_SUPERUSER_LIST - x,x,x
```

Thank you @angela-tran

thekaveman reviewed (changes requested):
I'm a little confused as to if this is totally done?

* It seems like `ADMIN` checks were removed but the setting is still around. 
* Locally I'm seeing [this test failing](https://github.com/cal-itp/benefits/blob/dev/tests/pytest/core/test_settings.py#L7), should probably be deleted if we aren't using this setting anymore.
* The code in `benefits/admin.py` is now duplicated in `benefits/core/admin.py`.
* There is still code that checks for the `DJANGO_ADMIN` environment variable, e.g. in `bin/init.sh` -- if we are now saying in this PR that admin is always on, that code should probably be cleaned up as well.

This PR doesn't feel quite ready / cleaned up.

machikoyasuda commented:
@thekaveman @angela-tran Sorry for the delay --- I removed the test, un-did the settings.py changes, removed extra admin file, updated the function call setting ( "benefits.core.admin.pre_login_user"  - has to add core in there ). And re-tested it all locally.

angela-tran reviewed (approved):
This looks good to me

thekaveman reviewed (approved):
Yep, me too! I went through all 4 scenarios again:

- [x] I can login with the `DJANGO_SUPERUSER_USERNAME` and password
- [x] I can login with my own `compiler.la` email configured in the superuser list
- [x] I can login with a different `compiler.la` email configured in the user list
- [x] I _cannot_ login with a personal Gmail email not in the `compiler.la` domain
